docs/src/spaces.md

```diff
@@ -0,0 +1,155 @@
+
+For a full listing of docstrings and methods, see the [Spaces Reference](@ref) section.
+
+## Usecases
+
+Good usecases for a Pkg scratchspace include:
+
+* Caching downloads of files that must be routinely accessed and modified by a package.  Files that must be modified are a bad fit for the immutable [Artifacts](@ref) abstraction, and files can always be re-downloaded if the cache is wiped by the user.
+
+* Generated data that depends on the characteristics of the host system.  Examples are compiled binaries, fontcache system font folder inspection output, generated CUDA bitcode files, etc...  Objects that would be difficult to compute off of the user's machine, and that can be recreated without user intervention are a great fit.
+
+* Directories that should be shared between multiple packages in a single depot.  The space keying mechanism (explained above) makes it simple to provide scratch space that can be shared between different versions of a package, or even between different packages.  This allows packages to provide a scratch space where other packages can easily find the generated data, however the typical race condition warnings apply here; always design your access patterns assuming another process could be reading or writing to this scratch space at any time.
+
+Bad usecases for a Pkg scratchspace include (but are not limited to):
+
+* Anything that requires user input to regenerate.  Because spaces can disappear, it is a bad experience for the user to need to answer questions at seemingly random times when the space must be rebuilt.
+
+* Storing data that is write-once, read-many times.  We suggest you use [Artifacts](@ref) for that, as they are much more persistent and are built to become portable (so that other machines do not have to generate the data, they can simple make use of the artifact by downloading it from a hosted location).  Spaces generally should follow a write-many read-many access pattern.
+
+## Frequently-Accessed Caching Questions
+
+> Can I trigger data regeneration if the space is found to be empty/files are missing?
+
+Yes, this is quite simple; just check the contents of the directory when you first call `get_space!()`, and if it's empty, run your generation function:
+
+```julia
+using Pkg, Pkg.Spaces
+
+function get_dataset_dir()
+    dataset_dir = @get_space!("dataset")
+    if isempty(readdir(dataset_dir))
+        perform_expensive_dataset_generation(dataset_dir)
+    end
+    return dataset_dir
+end
+```
+
+> Can I create a scratchspace that is not shared across versions of my package?
+
+Yes!  Make use of the `key` parameter and Pkg's ability to look up the current version of your package at compile-time:
+
+```julia
+module VersionSpecificExample
+using Pkg, Pkg.Spaces
+
+# Helpers to get current package UUID and VersionNumber
+function get_uuid()
+    return Base.PkgId(@__MODULE__).uuid
```

Comment:
Same with internal stuff here. I think we need an api that is not based on `PkgId`.